modules/tensor_mechanics/src/materials/ComputeEigenstrain.C

```diff
@@ -23,15 +25,21 @@ ComputeEigenstrain::validParams()
   return params;
 }
 
-ComputeEigenstrain::ComputeEigenstrain(const InputParameters & parameters)
-  : ComputeEigenstrainBase(parameters), _prefactor(getMaterialProperty<Real>("prefactor"))
+template <bool is_ad>
+ComputeEigenstrainTempl<is_ad>::ComputeEigenstrainTempl(const InputParameters & parameters)
+  : ComputeEigenstrainBaseTempl<is_ad>(parameters),
+    _prefactor(this->template getGenericMaterialProperty<Real, is_ad>("prefactor"))
 {
-  _eigen_base_tensor.fillFromInputVector(getParam<std::vector<Real>>("eigen_base"));
+  _eigen_base_tensor.fillFromInputVector(this->template getParam<std::vector<Real>>("eigen_base"));
 }
 
+template <bool is_ad>
 void
-ComputeEigenstrain::computeQpEigenstrain()
+ComputeEigenstrainTempl<is_ad>::computeQpEigenstrain()
 {
   // Define Eigenstrain
-  _eigenstrain[_qp] = _eigen_base_tensor * _prefactor[_qp];
+  this->_eigenstrain[this->_qp] = _eigen_base_tensor * _prefactor[this->_qp];
```

Comment:
@dschwen Do you know how to get rid of these annoying "this->" and the "this->template" above?

Reply:
With a `using` declaration. We'll basically have to bring back the `usingXXXMembers` macros that we had for the old AD classes that were templated on ComputeStage. It's not pretty, but it works.

Reply:
Okay I see. I guess I'll just forget about merging those AD materials in this PR and only put in the generic template (without testing?)